middle_end/flambda2/types/provers.ml

```diff
@@ -824,3 +824,211 @@ let never_holds_locally_allocated_values env var kind : _ proof_of_property =
+        if SS.is_empty (SS.inter s1 s2) then Proved false else Unknown
+      (* Immediates and allocated values -> Proved false *)
+      | ( Variant
+            { immediates = _;
+              blocks = Known blocks;
+              is_unique = _;
+              alloc_mode = _
+            },
+          ( Mutable_block _ | Boxed_float _ | Boxed_int32 _ | Boxed_int64 _
+          | Boxed_nativeint _ | Closures _ | String _ | Array _ ) )
+      | ( ( Mutable_block _ | Boxed_float _ | Boxed_int32 _ | Boxed_int64 _
+          | Boxed_nativeint _ | Closures _ | String _ | Array _ ),
+          Variant
+            { immediates = _;
+              blocks = Known blocks;
+              is_unique = _;
+              alloc_mode = _
+            } )
+        when TG.Row_like_for_blocks.is_bottom blocks ->
+        Proved false
+      (* Variants:
+       * incompatible immediates and incompatible block tags -> Proved false
+       * same immediate on both sides, no blocks -> Proved true
+       *)
+      | ( Variant
+            { immediates = immediates1;
+              blocks = blocks1;
+              is_unique = _;
+              alloc_mode = _
+            },
+          Variant
+            { immediates = immediates2;
+              blocks = blocks2;
+              is_unique = _;
+              alloc_mode = _
+            } ) -> (
+        match immediates1, immediates2, blocks1, blocks2 with
+        | Known imms, _, _, Known blocks
+          when TG.is_obviously_bottom imms
+               && TG.Row_like_for_blocks.is_bottom blocks ->
+          Proved false
+        | _, Known imms, Known blocks, _
+          when TG.is_obviously_bottom imms
+               && TG.Row_like_for_blocks.is_bottom blocks ->
+          Proved false
+        | Known imms1, Known imms2, Known blocks1, Known blocks2 -> (
+          let immediates_equality : _ generic_proof =
+            match
+              ( prove_naked_immediates_generic env imms1,
+                prove_naked_immediates_generic env imms2 )
+            with
+            | Invalid, _ | _, Invalid -> Invalid
+            | Unknown, _ | _, Unknown -> Unknown
+            | Proved imms1, Proved imms2 -> (
+              let module S = Targetint_31_63.Set in
+              if S.is_empty (S.inter imms1 imms2)
+              then Proved false
+              else
+                match S.get_singleton imms1, S.get_singleton imms2 with
+                | None, _ | _, None -> Unknown
+                | Some imm1, Some imm2 ->
+                  (* We've ruled out the empty intersection case, so the numbers
+                     have to be equal *)
+                  assert (Targetint_31_63.equal imm1 imm2);
+                  Proved true)
+          in
+          let blocks_equality : _ generic_proof =
+            let is_bottom1 = TG.Row_like_for_blocks.is_bottom blocks1 in
+            let is_bottom2 = TG.Row_like_for_blocks.is_bottom blocks2 in
+            if is_bottom1 && is_bottom2
```

Comment:
Hmm, should these four lines (942 to 945) match line 924?

Reply:
I think it the reverse: line 924 should match these lines. I will add a comment explaining why `Invalid` is given a slightly different semantics here.